oslo_messaging/metrics.py

```diff
@@ -0,0 +1,22 @@
+import logging
+import threading, queue
+import statsd
+
+LOG = logging.getLogger(__name__)
+
+class Metrics():
+   def __init__(self, host, port):
+      self._port = port
+      self.queue = queue.Queue()
+      self.statsd = statsd.StatsClient(host, port)
+
+   def start(self):
+      threading.Thread(target=self.counter, daemon=True).start()
+
+   def counter(self):
+      while True:
+         event = self.queue.get()
+         method = event.message.get('method')
+
+         stat = "oslo.messaging.{}.{}".format(method, event.message.tag)
+         self.statsd.incr(stat)
```

Comment:
There doesn't seem to be a way to gracefully stop this thread, which might lead to a container not shutting down in time, but hanging until timeout. Can we add such a way? Is there something in `oslo.messaging` that could call a stop method for this thread e.g. some method already existing to gracefully stop an rpc server?

Reply:
Ok we could add a stop method. But wont daemon threads automatically be killed once the main threads exits?

Reply:
possible stop event:
```
class Metrics():
   def __init__(self, host, port):
      self.stop_event = threading.Event()
      self.queue = queue.Queue()
      self.statsd = statsd.StatsClient(host, port)

   def start(self):
      threading.Thread(target=self.counter, daemon=True).start()

   def stop(self):
      self.stop_event.set()

   def counter(self):
      while not self.stop_event.is_set():
         event = self.queue.get()
         method = event.message.get('method')

         stat = "oslo.messaging.{}.{}".format(method, event.message.tag)
         self.statsd.incr(stat)
         self.queue.task_done()
```

Reply:
Yes. + `queue.get()` needs a timeout and then a `try-except` to catch [`Queue.Empty`](https://docs.python.org/2/library/queue.html#Queue.Empty) with a `continue` in the `except`

Reply:
With block=True and no timeout, it just waits forever until a new item is available. It should not throw an Empty Exception. (only when timeout is set and there was no new item within that time). Or is this not the same for python2 Queue?

Reply:
You're right, but we need a timeout of ~1s to let the `stop_event` have any effect. Otherwise, we would hang in the `queue.get()` anyways, as it's blocking, and would never check the condition in the while that would lead to a graceful shutdown.